ccm_web/server/src/canvas/canvas.interfaces.ts

```diff
@@ -142,33 +142,65 @@ export interface CourseWithSections extends CanvasCourseBase {
   sections: CanvasCourseSection[]
 }
 
-// Errors
+// Error Data
 
-interface CanvasError {
+export interface CanvasMessageErrorBody {
   message: string
 }
 
-function isCanvasError (value: unknown): value is CanvasError {
+export function isCanvasMessageErrorBody (value: unknown): value is CanvasMessageErrorBody {
   return hasKeys(value, ['message'])
 }
 
-export interface CanvasErrorBody {
-  errors: CanvasError[]
+export interface CanvasErrorsBody {
+  errors: unknown
 }
 
-export function isCanvasErrorBody (value: unknown): value is CanvasErrorBody {
-  if (!hasKeys(value, ['errors'])) {
-    return false
-  }
+export function isCanvasErrorsBody (value: unknown): value is CanvasErrorsBody {
+  return hasKeys(value, ['errors'])
+}
+
+export interface CanvasMessageErrorsBody extends CanvasErrorsBody {
+  errors: CanvasMessageErrorBody[]
+}
+
+export function isCanvasMessageErrorsBody (value: unknown): value is CanvasMessageErrorsBody {
+  if (!isCanvasErrorsBody(value)) return false
+  return (
+    Array.isArray(value.errors) &&
+    value.errors.every(e => isCanvasMessageErrorBody(e))
+  )
+}
 
-  if (Array.isArray(value.errors)) {
-    const result = value.errors.map(e => isCanvasError(e)).every(e => e)
-    return result
-  } else {
-    return true
+interface UniqueIdErrorData {
+  attribute: 'unique_id'
+  type: string
+  message: string
+}
+
+export interface CanvasUniqueIdErrorsBody extends CanvasErrorsBody {
+  errors: {
+    pseudonym: {
+      unique_id: UniqueIdErrorData[]
+    }
   }
 }
 
+export function isCanvasUniqueIdErrorsBody (value: unknown): value is CanvasUniqueIdErrorsBody {
+  if (!isCanvasErrorsBody(value)) return false
+  return (
+    hasKeys(value.errors, ['pseudonym']) &&
+    hasKeys(value.errors.pseudonym, ['unique_id']) &&
+    Array.isArray(value.errors.pseudonym.unique_id) &&
+    value.errors.pseudonym.unique_id.every(o => {
+      return (
+        hasKeys(o, ['attribute', 'type', 'message']) &&
+        Object.values(o).every(v => typeof v === 'string')
+      )
+    })
+  )
+}
```

Comment:
I'm not entirely sure this is necessary. I suppose we could just probe the object at runtime and check for `undefined`s and such. I go back and forth about how much to assume is `unknown` and what to type guard; it's an area I still am not super confident in, handling unknown data. Anyways, one benefit of this is we can also easily extract the message if the `type` doesn't happen `taken`.

Reply:
I understand.  I'm not sure it will always be the case that _**everything**_ inside `errors.pseudonym.unique_id` will have those keys.  So far, that's all we've seen, but I don't know what's possible.

Reply:
In case you want more background: https://www.typescriptlang.org/docs/handbook/2/types-from-types.html